### PR TITLE
feature(vrl): Add read-only support to `ExternalEnv`

### PR DESCRIPTION
**Pull request by fuchsnj**
closes https://github.com/vectordotdev/vector/issues/4744
adds ~90% of the changes needed for: https://github.com/vectordotdev/vector/issues/12731

changes:
- add VRL error code 315: "mutation of read-only value"
- add list of read-only paths to `ExternalEnv`. Paths are optionally recursive. There is a list for both the event, and metadata.
- check for read-only fields in `set_metadata_field` VRL function
- check for read-only fields in `remove_metadata_field` VRL function
- check for read-only fields in `del` VRL function
- check for read-only fields in assignment expressions
- add support for read-only paths to VRL tests, and add tests
- VRL conditions are now compiled as read-only, allowing the removal of an `Event` clone
- fix the `Look::starts_with` implementation for `LookupBuf`

## Example

config
```
[transforms.filter]
type = "filter"
inputs = [ "input" ]
condition.type = "vrl"
condition.source = """
. = "foo"
true
"""
```

error
```
error[E315]: mutation of read-only value
  ┌─ :1:1
  │
1 │ . = "foo"
  │ ^^^ mutation of read-only value
  │
  = see language documentation at https://vrl.dev
```

### CODE REVIEW

File: lib/vector-vrl-functions/src/remove_metadata_field.rs [hunk 1]
```diff
@@ -44,11 +44,21 @@ impl Function for RemoveMetadataField {
 
     fn compile(
         &self,
-        _state: (&mut state::LocalEnv, &mut state::ExternalEnv),
+        state: (&mut state::LocalEnv, &mut state::ExternalEnv),
```
Comment (pablosichert) [inline lib/vector-vrl-functions/src/remove_metadata_field.rs:47]: Is it possible to destructure here like

```suggestion
        (_, external_env): (&mut state::LocalEnv, &mut state::ExternalEnv),
```

so we don't have to read as `state.1` below?

File: lib/vrl/compiler/src/state.rs [hunk 2]
```diff
@@ -81,7 +98,63 @@ impl ExternalEnv {
...
+        }
+        false
+    }
+
+    /// Adds a path that is considered read only. Assignments to any paths that match
+    /// will fail at compile time
+    ///
+    /// # Panics
+    ///
+    /// Panics if the path contains coalescing.
```
Comment (pablosichert) [inline lib/vrl/compiler/src/state.rs:140]: Nit:

```suggestion
    /// Adds a path that is considered read only. Assignments to any paths that match
    /// will fail at compile time.
    ///
    /// # Panics
    ///
    /// Panics if the path contains coalescing.
```

File: lib/vrl/stdlib/src/del.rs [hunk 3]
```diff
@@ -83,12 +83,19 @@ impl Function for Del {
 
     fn compile(
         &self,
-        _state: (&mut state::LocalEnv, &mut state::ExternalEnv),
+        state: (&mut state::LocalEnv, &mut state::ExternalEnv),
```
Comment (pablosichert) [inline lib/vrl/stdlib/src/del.rs:86]: Same here regarding `state.1`

File: src/conditions/check_fields.rs [hunk 4]
```diff
@@ -555,24 +555,28 @@ pub struct CheckFields {
 }
 
 impl Conditional for CheckFields {
-    fn check(&self, e: &Event) -> bool {
-        self.predicates.iter().all(|(_, p)| p.check(e))
+    fn check(&self, e: Event) -> (bool, Event) {
+        let result = self.predicates.iter().all(|(_, p)| p.check(&e));
```
Comment (pablosichert) [inline src/conditions/check_fields.rs:559]: Should we give the `check` method on the predicate an owned `Event` as well?

**Review by pablosichert — APPROVED** [review]
Nice work, looks good!

Some small questions/comments:

File: src/conditions/check_fields.rs (hunk 4 above)
Comment (fuchsnj) [reply src/conditions/check_fields.rs:559]: I'd prefer to keep `&Event` where possible. VRL required an owned `Event`, but check field predicates don't use VRL.

**Pull request merged**